checksum checking for data.pkg and autoids

weanti opened the pull request.

weanti committed:
new message to distribute data integrity information via checksum

weanti committed:
Merge remote-tracking branch 'upstream/master'

weanti committed:
added file checksum calculator

weanti committed:
added autoid and data.pkg checksum distributing

weanti committed:
handle regular files as well as data.pkg content in filechecksum

weanti committed:
added checksum calculation and chekcum mismatch handling on client side
move checksum handing on server side: now it handles reconnects too

weanti committed:
indentation

weanti committed:
Merge remote-tracking branch 'upstream/master'

weanti committed:
remove unused include

MrPepperoni commented on src/platform/checksum.cpp:
```diff
@@ -0,0 +1,37 @@
+namespace platform {
+    boost::uint32_t fileChecksum( std::string const& filename )
+    {
+        boost::uint32_t checksum(0);
```
Probs its better to cache it, so each file is checksummed only once.

weanti replied:
on the server side it's calculated only once for each file (saved in static vars), on the client side it's supposed to receive once for each connecting.
do you see use cases when it maybe calculated several times?

MrPepperoni replied:
Scusi, my bad, didn't see the static

MrPepperoni commented:
Won't crlf vs lf make checksums fail? If so, a solution could be to check only select files inside data.pkg (autoids, map json files, actors), or better yet, r2pkg could convert text file lineendings before packaging.

weanti commented:
you're right. crlf <-> lf will cause problems

weanti committed:
Merge remote-tracking branch 'upstream/master'

weanti committed:
add EOL conversion for win <-> linux checksum interoperability

MrPepperoni commented on src/platform/checksum.cpp:
```diff
@@ -28,8 +28,26 @@ namespace platform {
         {
             f->ReadAll( data );
         }
+        std::string dataNoEol("");
```
Probs I wasn't clear, I'd suggest removing lineendings altogether. UNIX-like lineending is \n, mac os before 9 uses \r, after that, \n.
Also consider using the boost string algorithms, erase_all / replace_all  http://www.boost.org/doc/libs/1_56_0/doc/html/string_algo/usage.html#idp432318416

weanti replied:
you were clear, it's not final
I was looking for a good solution for removing both kind of EOLs without too much memory usage and iterating several times, so your suggestion is valuable

weanti committed:
remove special handling of data.pkg in fileChecksum
added EOL removing in fileChecksum

weanti committed:
add checksum getter for package
MISSING: checksum calculation and verification

weanti committed:
add file specific checksum calculation to my_name_message_handler (server side) and data_checksum_message (client) side

weanti committed:
added checksum calculation and verification to Package
minor changes: naming, moved EOL removal to function

weanti committed:
fix compiler errors
fix checksum calculation in package writing: checksum was calculated on the compressed stream instead of the noncompressed EOL free stream

weanti committed:
separate function for Package checksum calculation

weanti committed:
fix Checksum function: add missing return value
remove unneeded file seeking in checksum calculation

weanti committed:
use specific functon for Package checksumming, now value can be saved static on server side, so no extra checksumming is needed every client join

weanti committed:
remove debug message

weanti commented:
- checksum calculation for data.pkg: calculated at writing, verified by loading
- made checksum calculation (more or less) EOL independent:\n and \r are removed before checksumming
- client and server calculates checksum for autoids and data.pkg and negotiates the result

MrPepperoni commented on src/network/my_name_message_handler_sub_system.cpp:
```diff
@@ -128,8 +129,27 @@ void MyNameMessageHandlerSubSystem::Execute( Message const& message )
+    // calculate checksums and distribute it to the clients
+    {
+        static boost::uint32_t datapkgChecksum = PackageChecksum("data.pkg");
+        static boost::uint32_t autoidChecksum = FileChecksum("autoids");
```
there's no such file

weanti replied:
there is one under run/data. I used that for testing (by copying next to the executable)
did I misunderstand the requirement again and autoid checking should be done otherwise?
or autoids is part of data.pkg?

MrPepperoni commented on src/platform/package.cpp:
```diff
@@ -21,7 +23,7 @@ class PackageImpl
         uint32_t Magic;
         uint32_t Version;
-        uint32_t Checksum;
+        boost::uint32_t Checksum;
```
we don't use the boost prefix pretty much anywhere

weanti replied:
okay

MrPepperoni commented on src/platform/checksum.cpp:
```diff
@@ -0,0 +1,43 @@
+
+namespace platform {
+    // calculates checksum on the given file, to make it multiplatform EOL characters are ripped before checksum
+    boost::uint32_t FileChecksum( std::string const& filename )
```
I'd just add this to ifile.*; using readall, removeeol, and checksumming. Then it works for osfiles and packaged files as well.

weanti replied:
very good idea

MrPepperoni commented on src/platform/checksum.cpp:
```diff
@@ -0,0 +1,43 @@
+        boost::erase_all( data,"\r" );
+    }
+
+    boost::uint32_t PackageChecksum( std::string const& packagename )
```
probs an easier solution is to put this into filesystem.*, xor the checksums of all the packages, and compare that. No need to hardcode the data.pkg name anywhere that way.

MrPepperoni commented on src/platform/package.h:
```diff
@@ -2,7 +2,6 @@
 #define INCLUDED_PLATFORM_PACKAGE_H
 
 #include <boost/filesystem/path.hpp>
-#include <memory>
```
I'm pretty sure auto_ptr will complain, at least on some platforms

weanti committed:
integrating checksumming into File as a feature (partial implementation)

weanti committed:
move checksumming feature into File
remove autoids checksumming, because that's part of data.pkg
remove some unused includes

weanti committed:
remove checkumming from File, because it's not used anywhere and Package checksumming uses a different approach
add proper chscksum validation to Package

weanti committed:
remove checksum verification at header loading

weanti committed:
remove debug msg

weanti committed:
undelete include <memory> for auto_ptr

weanti committed:
Merge remote-tracking branch 'upstream/master'
